### PR TITLE
Blog app: add authorization rules

### PR DESCRIPTION
**Pull request by Tafloninno**
### **In this PR @ismailmunyentwari9  and @Tafloninno  completed the following **

- Install CanCanCan in your project.

- Add a role column to the users table. Remember to use a migration for this.

- A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
For that you need to implement the post deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

- A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
For that you need to implement the comment deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -7,12 +7,19 @@ class User < ApplicationRecord
...
   # before_validation :set_default_posts_counter
 
   def recent_posts
     posts.order(created_at: :desc).limit(3)
   end
 
+  def admin?
+    role == 'admin'
+  end
+
```
Comment (KDlamini) [inline app/models/user.rb:22]: Kudos for creating the `user.admin?` method inside the user model. Fantastic job!!! 👏🏾💯

File: app/views/posts/show.html.erb [hunk 2]
```diff
@@ -3,25 +3,33 @@
...
-            <%= form_with model: Like, url: user_post_likes_path(@posts.author, @posts) do |form| %>
+            <%= form_with model: Like, url: user_post_likes_path(@post.author, @post) do |form| %>
                 <%= form.submit 'Like Post' %>
             <% end %>
+            
+        </div>
+        <div class="button_holder">
+           <% if can? :destroy, @post%>
+        <%= button_to "Delete post", user_post_path(id: @post.id), method: :delete, class:"delete_button" %>
+        <% end %>
```
Comment (KDlamini) [inline app/views/posts/show.html.erb:23]: Great job using the `can? :destroy` method to authorize the destroy actions. 🥂🎆

File: app/views/posts/show.html.erb [hunk 3]
```diff
@@ -3,25 +3,33 @@
...
+                <h3 class="title"><%= @post.title %> #<%= @post.id%> by <%= @user.name %></h3>
+                <li class="countcomentlikes">Comments:  <%= @post.comments_counter %>, Likes: <%= @post.likes_counter %> </li>
             </div>
-            <li class="bio_content"> <%= @posts.text %> </li>
+            <li class="bio_content"> <%= @post.text %> </li>
         </ul>
         <button class="page-btn">
-            <%= link_to 'Add New Comment', new_user_post_comment_path(@posts.author, @posts), style: 'color: white; text-decoration: none;' %>
+            <%= link_to 'Add New Comment', new_user_post_comment_path(@post.author, @post), style: 'color: white; text-decoration: none;' %>
         </button>
```
Comment (KDlamini) [inline app/views/posts/show.html.erb:13]: I am unable to create a new comment on any post. Every time I click the `Add new comment` button nothing happens.

- Kindly look into the authorization rules or perhaps the comments create action and make sure that every user is authorized to comment on any post including posts they do not own.

**Review by KDlamini — CHANGES_REQUESTED** [review]
### Hi team @Tafloninno, @ismailmunyentwari9 🙋‍♀️,  

Great work so far! 💯 
You have done a great job adding the Blog app authorization rules. 🎊 🥳 
There are some issues that you still need to work on to improve your project according to the requirements. You are almost there! 💪🏾 

## Highlights 🌞 

✔️  Good PR title and description
✔️  Users are only authorized to delete posts and comments they own
✔️  The `role` column exists in the user's table
✔️  There's a CanCanCan `ability` class that defines the set of rules for user-post interactions
✔️  Project has **CanCan** Gem in **Gemfile**

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @KDlamini  in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/posts/show.html.erb (hunk 3 above)
Comment (Tafloninno) [reply app/views/posts/show.html.erb:13]: @KDlamini  Thanks for the review 
but whenever I click on add comment, it works on my side

Comment (Tafloninno) [reply app/views/posts/show.html.erb:13]: Screenshot 2023-06-23 at 15 59 31

**Review by AnselemOdims — APPROVED** [review]
### Hi  @Tafloninno 🙋‍♀️, 
Good job in making the required changes 👏 

# STATUS APPROVED ✔
Your project is complete!  There is nothing else to say other than... it's time to merge it 🔗 🚢 
Congratulations! 🎉 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers 🥂  and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Use the tag @AnselemOdims for a quicker reply


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**